### PR TITLE
[inductor] Fix inductor dropping ordering dep between effectful ops with different kernel types

### PR DESCRIPTION
**Pull request by pkourdis**
`with_effects` records an ordering edge between consecutive ORDERED-effectful ops so the scheduler preserves their program order. That edge was both dropped and, when it did apply, too strong. Fixing the strength then exposed a latent conflation in how pinned buffers are freed.

**The edge was dropped.** It was recorded in `V.graph.additional_star_deps` keyed by `new_op.get_name()` (a buffer name), but the scheduler reads it back by operation name. Ops lowering to the same kernel type gave names that coincided, masking the mismatch; different kernel types (e.g. an `ExternKernel` followed by a `_CollectiveKernel`) made the key miss, so the dep vanished and the scheduler could reorder `op_b` ahead of `op_a`. Fixed by keying on `get_operation_name()`.

**The edge was too strong.** `additional_star_deps` is installed as a `StarDep`, which asserts ordering *and* liveness, so an effect edge counted as a real read of the previous op's output buffer and held it live past its last true use, blocking reuse and deallocation. An effect edge carries no data, so it belongs in `additional_buffer_deps`, installed as `WeakDep(is_fake=True)`, filtered out of `used_or_aliased_buffer_names`, so it orders without extending lifetimes. This is the pattern `OrderingBarrier` already documents. `additional_star_deps` had no other users and is removed.

**Retention is a separate concern, and also needed.** The above is about the *previous* op's output buffer. Independently, *this* op's inputs may be retained in state inductor cannot see (`tq.push(x)` stashes a tensor in a torchbind queue), and no dependency edge can express "the callee kept a reference", so those inputs must not have their storage recycled. The axes are independent: without the pin `test_torchbind_queue` gives silently wrong numerics under either dep strength.

**Pinning must block recycling without blocking deallocation,** since freeing only drops inductor's own reference and storage retained behind its back survives. That distinction did not exist: `never_reuse_buffers` both withheld a buffer from the reuse pool and suppressed its free. Putting retention pins there would therefore keep every effectful op's input allocated for the whole graph, and peak memory grows with the number of effectful ops — measured locally, though the memory test is not part of this PR. But the suppression half is load-bearing elsewhere: `ShallowCopyDataKernel` pins its own output, which in the cpp wrapper has no output variable at all (the call is emitted bare), so emitting a free for it generates a `reset()` on an undeclared name and fails to compile. Rather than redefine the existing set, retention pins go in a new `never_reuse_but_free_buffers`, and the split is expressed as `can_free` (body unchanged from the old `can_reuse`, so every pre-existing pin site behaves exactly as before) plus `can_reuse` (`can_free` and not retention-pinned).

Review `lowering.py` first for the keying fix, the weakened dep, and the pin; then `scheduler.py` for the dep the switch now goes through, and `graph.py` / `wrapper.py` for the two sets and the `can_free`/`can_reuse` split.

## Test plan

`test_inductor_preserves_effect_order_across_kernel_types` covers the keying fix: two ORDERED ops lowering to different kernel types (`op_a` -> `FallbackKernel`, `op_b` -> `_CollectiveKernel`), asserting `op_a` is emitted first. It fails without the fix because the dropped dep lets the scheduler reorder them.

`test_inductor_effect_order_does_not_extend_lifetimes` covers the dep strength. To discriminate it needs an effectful op with a real output buffer for a strong dep to over-retain, plus a later effectful op consuming a value derived from that buffer so it cannot be hoisted above the buffer's last use; it then asserts the buffer is freed before that later op. Verified to fail with `is_fake=False` and pass with `is_fake=True`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by EikanWang — COMMENTED** [review]
LGTM.

**Comment by EikanWang** [comment]
@pkourdis , let's wait for the CI signal.

**Review by EikanWang — CHANGES_REQUESTED** [review]
@pkourdis , the failures should be related. Please check all these failures.

**Comment by EikanWang** [comment]
@claude are the failures related?

**Comment by pkourdis** [comment]
@EikanWang I am looking into the related failures and will follow up.

**Comment by pkourdis** [comment]
@EikanWang I have pushed an additional fix to mark effectful ops' tensor inputs as non-reusable so inductor doesn't recycle memory still held in hidden state (e.g. a torchbind queue), which was corrupting results. All reported failing tests are passing according to my local runs.

**Comment by EikanWang** [comment]
@claude, are the failures related?

**Review by EikanWang — DISMISSED** [review]

**Review by desertfire — DISMISSED** [review]

**Review by jansel — DISMISSED** [review]

**Comment by pkourdis** [comment]
@zou3519 could you please help review?

**Review by EikanWang — DISMISSED** [review]

**Comment by pkourdis** [comment]
@EikanWang could you please take a look and merge it if everything looks good?

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by pkourdis** [comment]
> address automated feedback from claude

Thanks - addressed all three:

- Testing: 
  - Fixed by passing `lib=lib` to `_register_effectful_op`, so the handle is stored on the scoped library and released on context exit instead of leaking onto the global registry.
  - Added regression test `test_inductor_pins_effectful_op_input_buffers` that fails when the pin is removed.
 - Perfomance:
   - Confirmed that the change is intentional and documented that is a intended tradeoff for correctness since such numerical accuracy bugs are very hard to root cause on the application level.

**Comment by sramakintel** [comment]
With the comments addressed, can we merge this PR if everything is OK? This PR merge is required to unblock CI hangs in torchcomms CI across all backends (e.g https://github.com/meta-pytorch/torchcomms/pull/3219)

**Comment by jansel** [comment]
@claude review these changes again

**Review by jansel — DISMISSED** [review]

**Comment by pkourdis** [comment]
@desertfire @EikanWang can you please review again and merge if everything looks good?

**Comment by pkourdis** [comment]
Can a maintainer please help to merge?

**Comment by sramakintel** [comment]
@EikanWang can we merge this PR?

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m="Reverted internally" -c="ghfirst"


The `never_reuse_buffers` pin this PR adds for ORDERED effectful-op inputs also prevents those buffers from being freed for the duration of the compiled call, so peak GPU memory scales with the number of effectful ops in the graph, causing training OOMs. See #192095 for the mechanism and a regression test. Please do not reland until #192095 lands and passes.

**Comment by eellison** [comment]
See, comment here: https://github.com/pytorch/pytorch/pull/183804#pullrequestreview-4566127760. 

We should be using WeakDep for ordering to prevent extending lifetimes.

**Comment by pkourdis** [comment]
> @pytorchbot revert -m="Reverted internally" -c="ghfirst"
> 
> The `never_reuse_buffers` pin this PR adds for ORDERED effectful-op inputs also prevents those buffers from being freed for the duration of the compiled call, so peak GPU memory scales with the number of effectful ops in the graph, causing training OOMs. See #192095 for the mechanism and a regression test. Please do not reland until #192095 lands and passes.

@izaitsevfb this PR has been updated and the (current) version of the test added in https://github.com/pytorch/pytorch/pull/192095 is now passing. Can you please expedite the land of https://github.com/pytorch/pytorch/pull/192095?

**Comment by izaitsevfb** [comment]
> @izaitsevfb this PR has been updated and the (current) version of the test added in #192095 is now passing. Can you please expedite the land of #192095?

the test PR is blocked on code review, I'm trying to ping the test author internally. will keep you posted, sorry that it takes so long

**Comment by sramakintel** [comment]
@izaitsevfb any updates on the merge of this PR? I have also followed up on https://github.com/pytorch/pytorch/pull/192095 on updates. can we merge this PR? all tests are passing.This will unblock the torchcomms CI that has been facing hang issues for a long time now

**Comment by izaitsevfb** [comment]
> @izaitsevfb any updates on the merge of this PR? I have also followed up on #192095 on updates. can we merge this PR? all tests are passing.This will unblock the torchcomms CI that has been facing hang issues for a long time now

@sramakintel. no updates unfortunately, looks like the internal workstream is stalled.  if this PR passes the test added in https://github.com/pytorch/pytorch/pull/192095 , let's try merging it again. 

sorry for the delay.

**Comment by pkourdis** [comment]
> > @izaitsevfb any updates on the merge of this PR? I have also followed up on #192095 on updates. can we merge this PR? all tests are passing.This will unblock the torchcomms CI that has been facing hang issues for a long time now
> 
> @sramakintel. no updates unfortunately, looks like the internal workstream is stalled. if this PR passes the test added in #192095 , let's try merging it again.
> 
> sorry for the delay.

@izaitsevfb l did run locally the latest version of the tests added in https://github.com/pytorch/pytorch/pull/192095 and are passing. @eellison status is on vacation. @jansel can you please help review again?

**Comment by sramakintel** [comment]
@jansel and @izaitsevfb can we merge this PR? Since this is required to fix CI hangs on the torchcomms side, we will need progress on merging this PR. Please help expedite!

**Comment by jansel** [comment]
@izaitsevfb should review/merge here -- I don't have context on the internal failure

**Review by izaitsevfb — APPROVED** [review]
